torch/onnx/symbolic_opset11.py

```diff
@@ -94,11 +94,10 @@ def _apply(fn):
 @symbolic_helper.parse_args("v", "f", "f")
 @_beartype.beartype
 def hardtanh(g: jit_utils.GraphContext, self: _C.Value, min_val: float, max_val: float):
-    dtype = self.type().scalarType()
-    if dtype is None:
+    try:
+        scalar_type = _type_utils.JitScalarType.from_value(self)
+    except errors.OnnxExporterError:
         scalar_type = _type_utils.JitScalarType.FLOAT
```

Comment:
Hopefully we can avoid try catches here as well. If we prefer it though, a ValueError seems to me a better choice.

Reply:
`try-except` is the only option if we don't want to return `None` from `JitScalarType.from_` APIs. It would be great to have discussed error handling and other aspects on a design document before https://github.com/pytorch/pytorch/pull/81953 was landed

Non custom exception as `ValueError` is not desired inside `torch.onnx`

Reply:
I don't know if we want to shy away from all python exceptions in torch.onnx. As a user I would expect to get OnnxExporterError when I use export, but internally I would expect a ValueError when I call a function with values it doesn't expect. Catching a ValueError and re-raising may work but sounds too cumbersome. Is there a reason why we don't want to return None?

As an alternative, is it possible to create an error that subclasses from OnnxExporterError for this scenario, so that we don't catch for `OnnxExporterError` in the caller, which is overly broad? Happy to hear other opinions and ideas too

Reply:
Sorry, I don't fully understand who is "as user" or what is "internally" without examples or more explanation.

Nonetheless, this exception is not central for the PR; it was added following a proposal discussed during one of our weekly meetings and [opened for discussion](https://github.com/pytorch/pytorch/issues/76155) with the actual use-case that triggered the topic (ATen Fallback confusing RuntimeError coming from symbolic functions and other places). We can work on the exception part in a separate PR. For now, I prefer not to create a new exception without taking into consideration corner cases, requirements, etc. The original exception issue is still opened, so feel free to add your thoughts and/or examples where native exceptions are useful within torch.onnx. Hopefully we can work on it soon

Reply:
Sorry should have clarified: Users=torch users and users who creates custom symbolics using torch.onnx apis; internally=code inside torch.onnx.

Giving it a second thought, I actually like the idea of rethrowing now. In most cases, we just catch and set a default, so nothing actually gets thrown to users.

In cases where we do want to throw, the `value` can be added to the error with the `raise xxx from xxx` syntax, and users will get a OnnxExporterError to handle, with context.

But I agree it is tricky. For example, if from_value throws in a place where we don't expect it to and thus do not have error handling code, a user will see ValueError, which is what we don't want like you mentioned.

Keeping this as is SGTM for this PR

Reply:
New thought: Can `from_value` take an optional default value, and when it does it doesn't throw? I feel like we can have the best behavior with this, and make the caller easier too.

Reply:
lol that was what the [PR proposed](https://github.com/pytorch/pytorch/pull/87245/commits/4b5c96a38c59d59a92332943c24a840f1b06ea87#) with `raises=True`. I've changed this on your [request](https://github.com/pytorch/pytorch/pull/87245#discussion_r1003575317). that is when I realized ` -> Union[JitScalarType, None]` and ` -> Optional[JitScalarType]` would cause a lot of trouble with CI (maybe a package version issue or something)

`allow_list_type` was also suppressed, but I think it makes sense to have it removed. I've added it specifically for scenarios using `is_in_type_group()` helper

I was really divided in which approach to follow, but returning None made callers life a lot easier, indeed.
I can add it back after your next round of review

Reply:
> Sorry should have clarified: Users=torch users and users who creates custom symbolics using torch.onnx apis; internally=code inside torch.onnx.
> 
> Giving it a second thought, I actually like the idea of rethrowing now. In most cases, we just catch and set a default, so nothing actually gets thrown to users.
> 
> In cases where we do want to throw, the `value` can be added to the error with the `raise xxx from xxx` syntax, and users will get a OnnxExporterError to handle, with context.
> 
> But I agree it is tricky. For example, if from_value throws in a place where we don't expect it to and thus do not have error handling code, a user will see ValueError, which is what we don't want like you mentioned.
> 
> Keeping this as is SGTM for this PR

cool, will add this comment to the discussion issue

Reply:
I think the new proposal is slightly different in that:

1. When a default is specified, the default is returned in a case where an error would be raised otherwise
2. When a default is not specified, the error is raised.

So no None will be returned.

I imagine usage can look like

```python
_type_utils.JitScalarType.from_value(self)  # Will raise if there is error

_type_utils.JitScalarType.from_value(self, default=_type_utils.JitScalarType.FLOAT)  # Will return FLOAT if there is error

_type_utils.JitScalarType.from_value(self, default=None)
```

etc.

Reply:
Very nice example, thank you!

That works well and it is pythonic in the sense that other popular apis use it, such as the [dict.get](https://python-reference.readthedocs.io/en/latest/docs/dict/get.html)

Will implement this change now